File: solutions/ruby/dnd-character/2/dnd_character.rb
=begin
Write your code for the 'D&D Character' exercise in this file. Make the tests in
`dnd_character_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/dnd-character` directory.
=end

class DndCharacter
  
  def self.modifier(constitution)
    ((constitution - 10) / 2).floor
  end

  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def initialize
    @strength     = self.roll
    @dexterity    = self.roll
    @constitution = self.roll
    @intelligence = self.roll
    @wisdom       = self.roll
    @charisma     = self.roll
    @hitpoints    = 10 + DndCharacter.modifier(@constitution)
  end

  private

  def roll
    results = []
    4.times { results.push(Random.random_number(1..6)) }
    results.sort[1..].sum
  end

end

File: solutions/lua/dnd-character/2/dnd-character.lua
local Character = {}

local function sum(...)
  local result = 0
  for _, v in ipairs({...}) do
    result = result + v
  end
  return result
end

local function ability()
  local rolls = {
    math.random(1, 6),
    math.random(1, 6),
    math.random(1, 6),
    math.random(1, 6)
  }
  table.sort(rolls)
  return sum(table.unpack(rolls, 1, #rolls - 1))
end

local function modifier(character_constitution)
  return math.floor((character_constitution - 10) / 2)
end

function Character:new(name)
  local character = {
    name = name,
    strength = ability(),
    dexterity = ability(),
    constitution = ability(),
    intelligence = ability(),
    wisdom = ability(),
    charisma = ability()
  }

  character['hitpoints'] = 10 + modifier(character.constitution)
  
  self.__index = self
  return setmetatable(character, self)
end

return {
  Character = Character,
  ability = ability,
  modifier = modifier
}
